rpc/RPCSession.java

```diff
@@ -65,66 +74,282 @@ public RPCSession(GraknClient client, String database, Type type, GraknOptions o
+            this.databaseReplicas = discoverDatabaseReplicas();
+            coreSessions = new ConcurrentHashMap<>();
+            isOpen =true;
+        }
+
+        @Override
+        public Grakn.Transaction transaction(Grakn.Transaction.Type type) {
+            return transaction(type, new GraknOptions());
+        }
+
+        @Override
+        public Grakn.Transaction transaction(Grakn.Transaction.Type type, GraknOptions options) {
+            DatabaseReplica selected = type != Grakn.Transaction.Type.READ_REPLICA ?
+                    databaseReplicas.selectLeader() : databaseReplicas.selectedReplica();
+            RPCSession.Core selection = coreSessions.computeIfAbsent(
+                    selected.id(),
+                    key -> {
+                        LOG.info("Opening a session to leader '{}'", selected);
+                        return clusterClient.coreClients().get(key.address()).session(key.database(), this.type, this.options);
+                    }
+            );
+            LOG.info("Opening a transaction of of type '{}' to leader '{}'", type, selected);
+            return selection.transaction(type, options);
+        }
+
+        @Override
+        public Type type() {
+            return type;
+        }
+
+        @Override
+        public boolean isOpen() {
+            return isOpen;
+        }
+
+        @Override
+        public void close() {
+            coreSessions.values().forEach(Core::close);
+            isOpen = false;
+        }
+
+        @Override
+        public String database() {
+            return database;
+        }
+
+        private DatabaseReplicas discoverDatabaseReplicas() {
+            int attempt = 0;
+            while (attempt < clusterClient.coreClients().size()) {
+                try {
+                    return DatabaseReplicas.ofProto(
+                            clusterDiscoveryRPC.databaseReplicasDiscovery(
+                                    DiscoveryProto.Discovery.DatabaseReplicas.Req.newBuilder()
+                                            .setDatabase(database)
+                                            .build()
+                            )
+                    );
+                } catch (StatusRuntimeException e) {
+                    ClusterGrpc.ClusterBlockingStub old = clusterDiscoveryRPC;
+                    clusterDiscoveryRPC = this.clusterClient.selectNextClusterDiscoveryRPC();
+                    LOG.info("Cluster Discovery RPC failed - server '{}' not available. Attempting to perform the operation onto another server '{}'",
+                            old.getChannel().authority(), clusterDiscoveryRPC.getChannel().authority());
+                }
+            }
+
+            throw new GraknClientException(CLUSTER_NOT_AVAILABLE.message(clusterClient.coreClients()));
+        }
+
+        public static class DatabaseReplicas {
+            private final ConcurrentMap<DatabaseReplica.Id, DatabaseReplica> replicaMap;
```

Comment:
Same comment as above - does this need to be a `ConcurrentMap` or would `HashMap` do?

Reply:
Correct. Fixed in dcb936cf1c33b48cb74540a621aad9becb505d34